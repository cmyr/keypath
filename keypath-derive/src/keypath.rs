use std::borrow::Borrow;
use std::fmt::Display;
use std::iter::FromIterator;

use proc_macro::token_stream::IntoIter as TokenIter;
use proc_macro::{Delimiter, Group, Ident, Literal, Punct, Spacing, Span, TokenStream, TokenTree};
use quote::{quote, quote_spanned};

use super::shared::PathComponent;

pub(crate) fn keypath_impl(input: TokenStream) -> Result<TokenStream, SyntaxError> {
    //eprintln!("{:?}", input);
    let mut iter = input.into_iter();
    let root: TokenStream = TokenTree::Ident(require_ident(&mut iter)?).into();
    let root: proc_macro2::TokenStream = root.into();
    let path_elements = collect_path_elements(&mut iter)?;
    let element_validators = path_elements.iter().map(PathElement::traverse_type);
    let element_fields = path_elements.iter().map(PathElement::to_tokens);
    let tokens = quote!(
        #root::get()
        #( .#element_validators )*

        .to_key_path_with_root::<#root>(&[#( #element_fields ),*])
    );
    //eprintln!("{}", tokens);
    Ok(tokens.into())
}

struct PathElement {
    element: PathComponent,
    span: Span,
}

impl PathElement {
    fn traverse_type(&self) -> proc_macro2::TokenStream {
        let ident = self.element.to_tokens();
        let span = self.span.into();
        quote_spanned!(span=> #ident)
    }

    fn to_tokens(&self) -> proc_macro2::TokenStream {
        let tokens = self.element.path_component_tokens();
        let span = self.span.into();
        quote_spanned!(span=> #tokens)
    }
}

pub(crate) struct SyntaxError {
    message: String,
    span: Span,
}

impl SyntaxError {
    pub(crate) fn into_compile_error(self) -> TokenStream {
        // compile_error! { $message }
        TokenStream::from_iter(vec![
            TokenTree::Ident(Ident::new("compile_error", self.span)),
            TokenTree::Punct({
                let mut punct = Punct::new('!', Spacing::Alone);
                punct.set_span(self.span);
                punct
            }),
            TokenTree::Group({
                let mut group = Group::new(Delimiter::Brace, {
                    TokenStream::from_iter(vec![TokenTree::Literal({
                        let mut string = Literal::string(&self.message);
                        string.set_span(self.span);
                        string
                    })])
                });
                group.set_span(self.span);
                group
            }),
        ])
    }
}

fn next_token(iter: &mut TokenIter) -> Result<TokenTree, SyntaxError> {
    iter.next().ok_or_else(|| SyntaxError {
        message: "unexpected end of input".to_owned(),
        span: Span::call_site(),
    })
}

fn syntax<T: Borrow<TokenTree>, M: Display>(token: T, message: M) -> SyntaxError {
    SyntaxError {
        message: message.to_string(),
        span: token.borrow().span(),
    }
}

fn require_ident(iter: &mut TokenIter) -> Result<Ident, SyntaxError> {
    match next_token(iter)? {
        TokenTree::Ident(ident) => Ok(ident),
        other => Err(syntax(other, "expected ident")),
    }
}

fn try_append_components(
    iter: &mut TokenIter,
    elements: &mut Vec<PathElement>,
) -> Result<(), SyntaxError> {
    let token = next_token(iter)?;
    let span = token.span();
    match &token {
        TokenTree::Literal(lit) => {
            match litrs::Literal::from(lit.clone()) {
                litrs::Literal::Integer(int) => {
                    let element = int
                        .value::<usize>()
                        .map(PathComponent::unnamed)
                        .ok_or_else(|| syntax(token, "indexes must be unsigned integers"))?;
                    elements.push(PathElement { element, span });
                }
                // a path like "This.hi.0.2" will have "0.2" parsed as a float literal
                litrs::Literal::Float(float)
                    if float.type_suffix().is_none() && float.exponent_part().is_empty() =>
                {
                    let first = float
                        .integer_part()
                        .parse::<usize>()
                        .map(PathComponent::unnamed)
                        .map_err(|_| syntax(&token, "indexes must be unsigned integers"))?;
                    let second = float
                        .fractional_part()
                        .ok_or_else(|| syntax(&token, "paths should not have trailing periods"))?
                        .parse::<usize>()
                        .map(PathComponent::unnamed)
                        .map_err(|_| syntax(token, "indexes must be unsigned integers"))?;
                    elements.push(PathElement {
                        element: first,
                        span,
                    });
                    elements.push(PathElement {
                        element: second,
                        span,
                    });
                }
                other => {
                    return Err(syntax(
                        token,
                        format!("expected index or field name, found literal '{}'", other),
                    ))
                }
            };
            Ok(())
        }
        TokenTree::Ident(ident) => {
            elements.push(PathElement {
                element: PathComponent::named(ident.to_string()),
                span,
            });
            Ok(())
        }
        other => Err(syntax(other, "expected ident")),
    }
}

fn parse_index(group: &Group) -> Result<PathElement, SyntaxError> {
    let mut group_tokens = group.stream().into_iter();
    //let lit = require_literal(&mut group_tokens)?;
    let token = next_token(&mut group_tokens)?;
    if let Some(token) = group_tokens.next() {
        return Err(syntax(token, "braces can only contain a single literal"));
    }
    let literal = match token.clone() {
        TokenTree::Literal(lit) => lit,
        _ => {
            return Err(syntax(
                &token,
                "keypath indexes must be string or integer literals",
            ))
        }
    };

    match litrs::Literal::from(literal) {
        litrs::Literal::Integer(int) => int
            .value::<usize>()
            .map(PathComponent::IndexInt)
            .ok_or_else(|| syntax(&token, "indexes must be unsigned integers")),
        litrs::Literal::String(s) => Ok(PathComponent::IndexStr(s.into_value().into_owned())),
        _ => Err(syntax(
            &token,
            "indexes may only be unsigned integers or strings",
        )),
    }
    .map(|element| PathElement {
        element,
        span: token.span(),
    })
}

/// Ok(true) when more work to do, Ok(false) when done
fn next_path_element(
    iter: &mut TokenIter,
    elements: &mut Vec<PathElement>,
) -> Result<bool, SyntaxError> {
    match iter.next() {
        None => return Ok(false),
        Some(TokenTree::Punct(p)) if p.as_char() == '.' && p.spacing() == Spacing::Alone => {
            try_append_components(iter, elements)?;
        }
        Some(TokenTree::Group(g)) if matches!(g.delimiter(), Delimiter::Bracket) => {
            let element = parse_index(&g)?;
            elements.push(element);
        }
        Some(token) => {
            eprintln!("BAD TOKEN {:?}", token);
            return Err(syntax(token, "expected '.' or '['"));
        }
    };

    //try_append_components(iter, elements)?;
    Ok(true)
}

fn collect_path_elements(iter: &mut TokenIter) -> Result<Vec<PathElement>, SyntaxError> {
    let mut result = Vec::new();
    loop {
        if !next_path_element(iter, &mut result)? {
            return Ok(result);
        }
    }
}
