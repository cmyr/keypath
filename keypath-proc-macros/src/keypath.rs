use proc_macro::TokenStream;
use quote::quote;

use super::keypath_parse::{KeyPathMacroInput, SyntaxError};

pub(crate) fn keypath_impl(input: TokenStream) -> Result<TokenStream, SyntaxError> {
    //eprintln!("{:#?}", input);
    let KeyPathMacroInput { root, components } = KeyPathMacroInput::parse(input)?;

    let element_validators = components.iter().map(|comp| comp.traverse_type());
    let element_fields = components.iter().map(|comp| comp.to_tokens());
    let tokens = quote!(
        <#root as ::keypath::Keyable>::Mirror::new()
        #( #element_validators )*

        .to_key_path_with_root::<#root>(&[#( #element_fields ),*])
    );
    //eprintln!("{}", tokens);
    Ok(tokens.into())
}
