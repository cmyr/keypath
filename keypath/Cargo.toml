[package]
name = "keypath"
version = "0.1.0"
authors = ["Colin Rofls <colin@cmyr.net>"]
edition = "2018"

[dependencies]
serde = "1.0"
serde_json = "1.0"
serde_derive = "1.0"
