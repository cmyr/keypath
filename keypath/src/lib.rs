#[macro_use]
extern crate serde_derive;
use std::any::Any;
use std::marker::PhantomData;

//mod lib2;
//mod json_keypath;
//mod lib3;
//mod newstart;
//mod newstart_index;
//mod value;

#[derive(Debug, Clone, Copy)]
struct SimplePath<T: 'static> {
    fields: &'static [Field],
    _type: PhantomData<T>,
}

impl<T: 'static> SimplePath<T> {
    fn new(fields: &'static [Field]) -> SimplePath<T> {
        SimplePath {
            fields,
            _type: PhantomData,
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub enum Field {
    Ord(usize),
    Name(&'static str),
}

trait RawKeyable: 'static {
    fn as_any(&self) -> &dyn Any;
    fn get_field(&self, ident: &[Field]) -> Result<&dyn RawKeyable, FieldError>;
}

trait Keyable: RawKeyable {
    fn item_at_path<T>(&self, path: &SimplePath<T>) -> Option<&T> {
        self.get_field(path.fields)
            .ok()
            .and_then(|t| t.as_any().downcast_ref())
    }
}

#[derive(Debug, Clone)]
enum FieldErrorKind {
    IndexOutOfRange(usize),
    InvalidField(Field),
}

#[derive(Debug, Clone)]
pub struct FieldError {
    kind: FieldErrorKind,
    type_name: &'static str,
    // the number of *remaining* fields at which the error occured
    depth: usize,
}

macro_rules! keyable_leaf {
    ($name:ty) => {
        impl RawKeyable for $name {
            fn as_any(&self) -> &dyn Any {
                self
            }

            fn get_field(&self, ident: &[Field]) -> Result<&dyn RawKeyable, FieldError> {
                match ident.split_first() {
                    None => Ok(self),
                    Some((head, rest)) => Err(FieldError {
                        kind: FieldErrorKind::InvalidField(head.to_owned()),
                        type_name: std::any::type_name::<Self>(),
                        depth: rest.len(),
                    }),
                }
            }
        }
    };
}

keyable_leaf!(String);
keyable_leaf!(f64);

impl<T: RawKeyable> RawKeyable for Vec<T> {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn get_field(&self, ident: &[Field]) -> Result<&dyn RawKeyable, FieldError> {
        match ident.split_first() {
            None => Ok(self),
            Some((Field::Ord(idx), rest)) => self
                .get(*idx)
                .ok_or_else(|| FieldError {
                    kind: FieldErrorKind::IndexOutOfRange(*idx),
                    type_name: std::any::type_name::<Self>(),
                    depth: rest.len(),
                })
                .and_then(|t| t.get_field(rest)),
            Some((field, rest)) => Err(FieldError {
                kind: FieldErrorKind::InvalidField(field.clone()),
                type_name: std::any::type_name::<Self>(),
                depth: rest.len(),
            }),
        }
    }
}

struct DemoStruct {
    friends: Vec<DemoPerson>,
}

struct DemoPerson {
    name: String,
    magnitude: f64,
}

impl RawKeyable for DemoPerson {
    fn get_field(&self, ident: &[Field]) -> Result<&dyn RawKeyable, FieldError> {
        match ident.split_first() {
            None => Ok(self),
            Some((Field::Name("name"), rest)) => self.name.get_field(rest),
            Some((Field::Name("magnitude"), rest)) => self.magnitude.get_field(rest),
            Some((field, rest)) => Err(FieldError {
                kind: FieldErrorKind::InvalidField(field.clone()),
                type_name: std::any::type_name::<Self>(),
                depth: rest.len(),
            }),
        }
    }

    fn as_any(&self) -> &(dyn Any + 'static) {
        self
    }
}

impl Keyable for DemoPerson {}

impl RawKeyable for DemoStruct {
    fn get_field(&self, ident: &[Field]) -> Result<&dyn RawKeyable, FieldError> {
        match ident.split_first() {
            None => Ok(self),
            Some((Field::Name("friends"), rest)) => self.friends.get_field(rest),
            Some((field, rest)) => Err(FieldError {
                kind: FieldErrorKind::InvalidField(field.clone()),
                type_name: std::any::type_name::<Self>(),
                depth: rest.len(),
            }),
        }
    }

    fn as_any(&self) -> &(dyn Any + 'static) {
        self
    }
}

impl Keyable for DemoStruct {}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn simple_keypath() {
        let mut person = DemoPerson {
            name: "Jojobell".to_string(),
            magnitude: 42.0,
        };
        let name_path = SimplePath::<String>::new(&[Field::Name("name")]);
        assert_eq!(person.item_at_path(&name_path).unwrap(), "Jojobell");
        person.name = "Colin".into();
        assert_eq!(person.item_at_path(&name_path).unwrap(), "Colin");
    }

    #[test]
    fn nested_keypath() {
        let person = DemoPerson {
            name: "coco".to_string(),
            magnitude: 42.0,
        };

        let person1 = DemoPerson {
            name: "jojo".to_string(),
            magnitude: 69.0,
        };

        let demo = DemoStruct {
            friends: vec![person, person1],
        };

        let jojo_name: SimplePath<String> =
            SimplePath::new(&[Field::Name("friends"), Field::Ord(1), Field::Name("name")]);

        assert_eq!(demo.item_at_path(&jojo_name).unwrap(), "jojo");
    }
}
