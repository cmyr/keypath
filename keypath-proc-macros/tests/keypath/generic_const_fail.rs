use keypath::{keypath, KeyPath, Keyable};

#[derive(Keyable)]
struct Container<T> {
    names: Vec<T>,
}

const _PATH: KeyPath<Container<String>, String> = keypath!(Container<String>.names[0]);

fn main() {
}
