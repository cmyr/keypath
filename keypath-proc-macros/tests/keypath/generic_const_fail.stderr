error[E0015]: calls in constants are limited to constant functions, tuple structs and tuple variants
 --> $DIR/generic_const_fail.rs:8:51
  |
8 | const _PATH: KeyPath<Container<String>, String> = keypath!(Container<String>.names[0]);
  |                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
